types/src/language_storage.rs

```diff
@@ -127,29 +138,81 @@ impl CryptoHash for ModuleId {
     }
 }
 
+impl CanonicalSerialize for PrimitiveTypeTag {
+    fn serialize(&self, serializer: &mut impl CanonicalSerializer) -> Result<()> {
+        use PrimitiveTypeTag::*;
+
+        match self {
+            Bool => serializer.encode_u8(0x0)?,
+            U64 => serializer.encode_u8(0x1)?,
+            ByteArray => serializer.encode_u8(0x2)?,
+            Address => serializer.encode_u8(0x3)?,
+        };
+
+        Ok(())
+    }
+}
+
+impl CanonicalDeserialize for PrimitiveTypeTag {
+    fn deserialize(deserializer: &mut impl CanonicalDeserializer) -> Result<Self> {
+        use PrimitiveTypeTag::*;
+
+        match deserializer.decode_u8()? {
+            0x0 => Ok(Bool),
+            0x1 => Ok(U64),
+            0x2 => Ok(ByteArray),
+            0x3 => Ok(Address),
+            _ => bail!("Can't deserialize tag fot PrimitiveTypeTag"),
+        }
+    }
+}
+
 impl CanonicalSerialize for StructTag {
     fn serialize(&self, serializer: &mut impl CanonicalSerializer) -> Result<()> {
-        serializer
-            .encode_struct(&self.address)?
-            .encode_struct(&self.module)?
-            .encode_struct(&self.name)?
-            .encode_vec(&self.type_params)?;
+        use StructTag::*;
+
+        match self {
+            Primitive(tag) => {
+                serializer.encode_u8(0x0)?.encode_struct(&tag)?;
+            }
+            Struct {
+                address,
+                module,
+                name,
+                type_params,
+            } => {
+                serializer
+                    .encode_u8(0x1)?
```

Comment:
I think extending `StructTag` to encompass primitives and structs may not be the right way to go here. The problems I see are:
- we use `StructTag` in many places besides `ContractEvent` (e.g., access paths for published resources), and primitive types are not possible in any of those contexts. Seems good to keep the type as tight as possible.
- This an extra byte to those other uses of `StructTag`

I think it might be preferable to keep `StructTag` as-is and add something called `TypeTag` that is used by events?

Reply:
Sounds reasonable.

Reply:
I think this is tied to the story of whether we want to be able to publish a resource with a generic type. If the answer is yes, we probably need to replace `StructTag` with `TypeTag` eventually.